cpp/oneapi/dal/backend/primitives/optimizers/cg_solver_dpc.cpp

```diff
@@ -0,0 +1,203 @@
+    return queue.submit([&](sycl::handler& cgh) {
+        cgh.memcpy(res_host, res_gpu, sizeof(Float));
+    });
+}
+
+template <typename Float, typename MatrixOperator>
+sycl::event cg_solve(sycl::queue& queue,
+                     MatrixOperator& mul_operator,
+                     const ndview<Float, 1>& b,
+                     ndview<Float, 1>& x,
+                     ndview<Float, 1>& residual,
+                     ndview<Float, 1>& conj_vector,
+                     ndview<Float, 1>& buffer,
+                     const Float tol,
+                     const Float atol,
+                     const std::int32_t maxiter,
+                     const event_vector& deps) {
+    // Solving the equation mul_operator(x) = b
+    const std::int64_t p = b.get_dimension(0);
+    ONEDAL_ASSERT(b.has_data());
+    ONEDAL_ASSERT(x.has_mutable_data());
+    ONEDAL_ASSERT(residual.has_mutable_data());
+    ONEDAL_ASSERT(conj_vector.has_mutable_data());
+    ONEDAL_ASSERT(buffer.has_mutable_data());
+    ONEDAL_ASSERT(x.get_dimension(0) == p);
+    ONEDAL_ASSERT(residual.get_dimension(0) == p);
+    ONEDAL_ASSERT(conj_vector.get_dimension(0) == p);
+    ONEDAL_ASSERT(buffer.get_dimension(0) == p);
+
+    Float alpha = 0, beta = 0;
+    Float r_norm = 0, b_norm = 0;
+    Float r_l1_norm = 0;
+
+    const auto kernel_minus = [=](const Float a, const Float b) -> Float {
+        return a - b;
+    };
+
+    auto compute_ax0_event = mul_operator(x, residual, deps); // r = Ax0
+
+    auto compute_r0_event = element_wise(queue,
+                                         kernel_minus,
+                                         residual,
+                                         b,
+                                         residual,
+                                         { compute_ax0_event }); // r0 = Ax0 - b
+    // compute_r0_event.wait_and_throw();
+    auto tmp_gpu = ndarray<Float, 1>::empty(queue, { 1 }, sycl::usm::alloc::device);
+    auto tmp_ptr = tmp_gpu.get_mutable_data();
+
+    l1_norm<Float>(queue, b, tmp_ptr, &b_norm, deps).wait_and_throw(); // compute norm(b)
+
+    // Tolerances for convergence norm(residual) <= max(tol*norm(b), atol)
+    Float threshold = std::max(tol * b_norm, atol);
+
+    const auto init_conj_kernel = [=](const Float residual_val, const Float conj_val) -> Float {
+        return -residual_val;
+    };
+    auto compute_conj_event = element_wise(queue,
+                                           init_conj_kernel,
+                                           residual,
+                                           conj_vector,
+                                           conj_vector,
+                                           { compute_r0_event }); // p0 = -r0 + 0 * p
+    auto conj_host = conj_vector.to_host(queue, {});
+    dot_product<Float>(queue, residual, residual, tmp_ptr, &r_norm, { compute_r0_event })
+        .wait_and_throw(); // compute r^T r
+
+    for (std::int32_t iter_num = 0; iter_num < maxiter; ++iter_num) {
```

Comment:
1) Please use `std::int64_t`
2) Please consider creation of `sycl::event last = ...` and assignment on each iteration